index.js

```diff
@@ -724,237 +495,227 @@ class XiaomiRoborockVacuum {
+
+        } else {
+          await this.device.activateCleaning();
+          this.log.info(`ACT setCleaning | ${this.model} | Start full cleaning, the device is in state ${this.device.property('state')}`);
+        }
+        return;
+      } else if (!state && (this.isCleaning || this.isInPause)) { // Stop cleaning
         await this.activateCharging(); // Charging works for 1st, not for 2nd
+        this.log.info(`ACT setCleaning | ${this.model} | Stop cleaning and go to charge, the device is in state ${this.device.property('state')}`);
+        //All rooms are removed from the list
+        this.roomsToClean.length = 0;
+        return;
       }
     } catch (err) {
-      this.log.error(
-        `ERR setCleaning | ${this.model} | Failed to set cleaning to ${state}`,
-        err
-      );
+      this.log.error(`ERR setCleaning | ${this.model} | Failed to set cleaning to ${state}`, err);
       throw err;
     }
+    this.log.error(`ERR setCleaning | ${this.model} |Cannot set cleaning to ${state} when current state is ${this.device.property('state')}!`);
+   // throw new Error('Cannot start or stop cleaning due to the current state!');
+
+  }
+  //Changed by Mai
+  async getCleaningRoom(room) {
+    await this.ensureDevice('getCleaningRoom');
+    return this.roomsToClean.includes(room);
   }
 
+  //Changed by Mai
   async setCleaningRoom(state, room) {
-    await this.ensureDevice("setCleaning");
+    await this.ensureDevice('setCleaning');
 
     try {
-      if (state && !this.isCleaning) {
-        // Start cleaning
-        this.log.info(
-          `ACT setCleaning | ${this.model} | Start cleaning Room ID ${room}, not charging.`
-        );
+      if (state && !this.isCleaning && !this.isInPause) {
+        this.log.info(`ACT setCleaning | ${this.model} | Add room ${room} to the list of rooms to be cleaned.`);
+        if (!this.roomsToClean.includes(room)) {
+          this.roomsToClean.push(room);
+        }
+        return;
+      } else if (!state && !this.isCleaning && !this.isInPause) {
+        this.log.info(`ACT setCleaning | ${this.model} | Remove room ${room} to the list of rooms to be cleaned.`);
+        if (this.roomsToClean.includes(room)) {
+          const index = this.roomsToClean.indexOf(room);
+          if (index > -1) {
+            this.roomsToClean.splice(index, 1);
+          }
+        }
+        return;
+      }
+    } catch (err) {
+      this.log.error(`ERR setCleaning | ${this.model} | Failed to set cleaning to ${state} for room ${room}`, err);
+      throw err;
+    }
+    this.log.warn(`WRN setCleaning | ${this.model} | Cannot set cleaning state ${state} for room ${room} when the robot is in state ${this.device.property('state')}`);
+    throw new Error(`WRN setCleaning | ${this.model} | Cannot set cleaning state ${state} for room ${room} when the robot is in state ${this.device.property('state')}`);
+  }
```

Comment:
Actual change to include in a PR

Reply:
That's the logic to add remove room id from the array.